src/main/java/de/tum/cit/aet/artemis/exercise/service/SubmissionFilterService.java

```diff
@@ -0,0 +1,138 @@
+ * required for the different types of exercises.
+ */
+@Service
+@Profile(PROFILE_CORE)
+public class SubmissionFilterService {
+
+    /**
+     * Filters a set of submissions from a participation and determines the latest submission with a valid result, based
+     * on the type of the exercise. This method is used to get data for the course dashboard, in which case the
+     * assessment due date matters, but also to get the latest results for each participation with points for each
+     * grading criterion, where the assessment due date can be ignored, because the corresponding endpoint is only
+     * visible for instructors.
+     *
+     * @param submissions             the submissions to filter
+     * @param ignoreAssessmentDueDate if the assessment due date should be ignored
+     * @return an optional containing the latest submission with a valid result if it exists, otherwise an empty optional
+     */
+    public Optional<Submission> getLatestSubmissionWithResult(Set<Submission> submissions, boolean ignoreAssessmentDueDate) {
+        if (submissions == null || submissions.isEmpty()) {
+            return Optional.empty();
+        }
+        return submissions.stream().filter(submission -> isSubmissionRelevantForCourseDashboard(submission, ignoreAssessmentDueDate)).max(Comparator.naturalOrder());
+    }
+
+    /**
+     * Determines if a submission is relevant for the course dashboard, and is safe to display there.
+     *
+     * @param submission the submission that is being checked.
+     * @return true if the submission can be displayed, false otherwise.
+     */
+    private boolean isSubmissionRelevantForCourseDashboard(Submission submission, boolean ignoreAssessmentDueDate) {
+        return switch (submission) {
+            case ProgrammingSubmission programmingSubmission -> isProgrammingSubmissionRelevantForCourseDashboard(programmingSubmission, ignoreAssessmentDueDate);
+            case ModelingSubmission modelingSubmission -> isModelingSubmissionRelevantForCourseDashboard(modelingSubmission, ignoreAssessmentDueDate);
+            case TextSubmission textSubmission -> isTextSubmissionRelevantForCourseDashboard(textSubmission, ignoreAssessmentDueDate);
+            case FileUploadSubmission fileUploadSubmission -> isFileUploadSubmissionRelevantForCourseDashboard(fileUploadSubmission, ignoreAssessmentDueDate);
+            case QuizSubmission quizSubmission -> isQuizSubmissionRelevantForCourseDashboard(quizSubmission);
+            default -> throw new IllegalArgumentException("Unsupported submission type: " + submission.getClass());
+        };
+    }
+
+    private boolean isProgrammingSubmissionRelevantForCourseDashboard(ProgrammingSubmission programmingSubmission, boolean ignoreAssessmentDueDate) {
+        var latestResult = programmingSubmission.getLatestResult();
+        var participation = (StudentParticipation) programmingSubmission.getParticipation();
+        var exercise = participation.getExercise();
```

Comment:
```suggestion
        Result latestResult = programmingSubmission.getLatestResult();
        StudentParticipation participation = (StudentParticipation) programmingSubmission.getParticipation();
        Exercise exercise = participation.getExercise();
```
* Artemis guidelines, if I remember correctly: only use `var` for complex types.
* Can we be sure this is a `StudentParticipation` here? The `public` entrypoint takes in an arbitrary `Set<Submission>`, some of which might be for the template or the solution.

Reply:
Good point. The method still works with the generic `Participation` type, so I just removed the cast just now.